temporal-sdk/src/main/java/io/temporal/internal/worker/AsyncActivityPollTask.java

```diff
@@ -0,0 +1,141 @@
+import com.google.protobuf.Timestamp;
+import com.uber.m3.tally.Scope;
+import io.grpc.Context;
+import io.temporal.api.common.v1.WorkerVersionCapabilities;
+import io.temporal.api.taskqueue.v1.TaskQueue;
+import io.temporal.api.taskqueue.v1.TaskQueueMetadata;
+import io.temporal.api.workflowservice.v1.GetSystemInfoResponse;
+import io.temporal.api.workflowservice.v1.PollActivityTaskQueueRequest;
+import io.temporal.api.workflowservice.v1.PollActivityTaskQueueResponse;
+import io.temporal.internal.common.GrpcUtils;
+import io.temporal.internal.common.ProtobufTimeUtils;
+import io.temporal.serviceclient.MetricsTag;
+import io.temporal.serviceclient.WorkflowServiceStubs;
+import io.temporal.worker.MetricsType;
+import io.temporal.worker.PollerTypeMetricsTag;
+import io.temporal.worker.tuning.ActivitySlotInfo;
+import io.temporal.worker.tuning.SlotPermit;
+import io.temporal.worker.tuning.SlotReleaseReason;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AsyncActivityPollTask implements AsyncPoller.PollTaskAsync<ActivityTask> {
+  private static final Logger log = LoggerFactory.getLogger(AsyncActivityPollTask.class);
+
+  private final TrackingSlotSupplier<?> slotSupplier;
+  private final WorkflowServiceStubs service;
+  private final Scope metricsScope;
+  private final PollActivityTaskQueueRequest pollRequest;
+  private final AtomicInteger pollGauge = new AtomicInteger();
+  private final Context.CancellableContext grpcContext = Context.ROOT.withCancellation();
+
+  @SuppressWarnings("deprecation")
+  public AsyncActivityPollTask(
+      @Nonnull WorkflowServiceStubs service,
+      @Nonnull String namespace,
+      @Nonnull String taskQueue,
+      @Nonnull String identity,
+      @Nullable String buildId,
+      boolean useBuildIdForVersioning,
+      double activitiesPerSecond,
+      @Nonnull TrackingSlotSupplier<ActivitySlotInfo> slotSupplier,
+      @Nonnull Scope metricsScope,
+      @Nonnull Supplier<GetSystemInfoResponse.Capabilities> serverCapabilities) {
+    this.service = service;
+    this.slotSupplier = slotSupplier;
+    this.metricsScope = metricsScope;
+
+    PollActivityTaskQueueRequest.Builder pollRequest =
+        PollActivityTaskQueueRequest.newBuilder()
+            .setNamespace(namespace)
+            .setIdentity(identity)
+            .setTaskQueue(TaskQueue.newBuilder().setName(taskQueue));
+    if (activitiesPerSecond > 0) {
+      pollRequest.setTaskQueueMetadata(
+          TaskQueueMetadata.newBuilder()
+              .setMaxTasksPerSecond(DoubleValue.newBuilder().setValue(activitiesPerSecond).build())
+              .build());
+    }
+
+    if (serverCapabilities.get().getBuildIdBasedVersioning()) {
+      pollRequest.setWorkerVersionCapabilities(
```

Comment:
You'll want to set deploymentOptions as well here, it applies to activities too not just WFTs. The sync poller might need updating as well if it doesn't have it.

In fact probably wouldn't hurt to de-dupe this stuff across the sync/async pollers

Reply:
OK, here's the item: https://github.com/temporalio/sdk-java/issues/2550